src/ucs/memory/rcache_vfs.c

```diff
@@ -0,0 +1,132 @@
+/**
+ * Copyright (C) 2021 NVIDIA CORPORATION & AFFILIATES. ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <ucs/config/global_opts.h>
+#include <ucs/datastruct/queue.h>
+#include <ucs/sys/compiler.h>
+#include <ucs/sys/math.h>
+#include <ucs/sys/string.h>
+#include <ucs/vfs/base/vfs_cb.h>
+#include <ucs/vfs/base/vfs_obj.h>
+#include "rcache_int.h"
+#include "rcache_vfs.h"
+
+
+#define UCS_RCACHE_FORMAT_STRING ("%12s:%9zi\n")
+
+
+static void ucs_rcache_vfs_read_inv_q_length(void *obj,
+                                             ucs_string_buffer_t *strb,
+                                             void *arg_ptr, uint64_t arg_u64)
+{
+    ucs_rcache_t *rcache = obj;
+    size_t rcache_inv_q_length;
+
+    ucs_spin_lock(&rcache->lock);
+    rcache_inv_q_length = ucs_queue_length(&rcache->inv_q);
+    ucs_spin_unlock(&rcache->lock);
+
+    ucs_string_buffer_appendf(strb, "%zu\n", rcache_inv_q_length);
+}
+
+static void ucs_rcache_vfs_read_gc_list_length(void *obj,
+                                               ucs_string_buffer_t *strb,
+                                               void *arg_ptr, uint64_t arg_u64)
+{
+    ucs_rcache_t *rcache = obj;
+    unsigned long rcache_gc_list_length;
+
+    ucs_spin_lock(&rcache->lock);
+    rcache_gc_list_length = ucs_list_length(&rcache->gc_list);
+    ucs_spin_unlock(&rcache->lock);
+
+    ucs_string_buffer_appendf(strb, "%lu\n", rcache_gc_list_length);
+}
+
+static void ucs_rcache_vfs_calc_regions_distribution(ucs_rcache_t *rcache,
+                                                     size_t *rcache_dist,
+                                                     size_t num_bins)
+{
+    ucs_config_thresholds_t *thresholds = &ucs_global_opts.rcache_thresholds;
+    ucs_rcache_region_t *region;
+    size_t region_size, i;
+
+    for (i = 0; i < num_bins; ++i) {
+        rcache_dist[i] = 0;
+    }
+
+    ucs_spin_lock(&rcache->lru.lock);
+
+    ucs_list_for_each(region, &rcache->lru.list, lru_list) {
+        region_size = region->super.end - region->super.start;
+
+        for (i = 0; i < num_bins; ++i) {
+            if (i == num_bins - 1) {
+                ++rcache_dist[i];
+            } else if (region_size < thresholds->vals[i]) {
+                ++rcache_dist[i];
+                break;
+            }
```

Comment:
looks like we can iterate up to `num_bins -1` and increase `rcache[num_bins -1]` unconditionally